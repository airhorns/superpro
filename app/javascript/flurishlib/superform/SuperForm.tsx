import { get, set, toPath } from "lodash";
import React from "react";
import Automerge from "automerge";
import { FieldPath, DocType } from "./utils";
import { ArrayHelpers } from "./ArrayHelpers";
import { SuperFormContext } from ".";

export type Command<T extends DocType> = (doc: T) => void;

export interface SuperFormProps<T extends DocType> {
  children: (form: SuperForm<T>) => React.ReactNode;
  initialValues?: T;
  onChange?: (doc: T, form: SuperForm<T>) => void;
  onSubmit?: (doc: T, form: SuperForm<T>) => void;
}

export type SuperFormErrors<T extends DocType> = {
  [K in keyof T]?: T[K] extends object ? SuperFormErrors<T[K]> : string;
};

interface SuperFormState<T extends DocType> {
  doc: T;
  errors: SuperFormErrors<T>;
}

export class SuperForm<T extends DocType> extends React.Component<SuperFormProps<T>, SuperFormState<T>> {
  currentBatch: Command<T>[] | null = null;

  constructor(props: SuperFormProps<T>) {
    super(props);
    let doc = Automerge.init<T>();

    if (this.props.initialValues) {
      doc = Automerge.change(doc, doc => Object.assign(doc, this.props.initialValues));
      // reset undo/redo state
      doc = Automerge.load(Automerge.save(doc));
    }

    this.state = { doc, errors: {} };
  }

  get doc(): T {
    return this.state.doc;
  }

  objectId(object: any) {
    return Automerge.getObjectId(object);
  }

  getValue(path: FieldPath) {
    return get(this.state.doc, path);
  }

  setValue(path: FieldPath, value: any) {
    this.dispatch(doc => set(doc, path, value));
  }

  deletePath(path: FieldPath) {
    const keys = toPath(path);
    this.dispatch(doc => {
      const root = get(doc, keys.slice(0, keys.length - 1));
      delete root[keys[keys.length - 1]];
    });
  }

  dispatch = (command: Command<T>) => {
    // Queue up changes for the batch if batching is underway
    if (this.currentBatch) {
      this.currentBatch.push(command);
      return;
    }

    // Otherwise, setState the result of applying the passed command to the document using Automerge
    const prevDoc = this.state.doc;
    this.setState(
      prevState => {
        const newDoc = Automerge.change(prevState.doc, command);
        return { doc: newDoc };
      },
      () => {
        console.debug("doc", this.state.doc);
        console.debug("changes", Automerge.getChanges(prevDoc, this.state.doc));
        this.props.onChange && this.props.onChange(this.state.doc, this);
      }
    );
  };

  batch = (callback: () => void) => {
    if (this.currentBatch) {
      // This is probably possible but it likely indicates the calling code isn't factored right, nested batches adds overhead and
      // commands should operate either at the batch level or atomic level, and batch level ones should only call atomic ones.
      throw new Error("can't nest superform batches");
    }

    this.currentBatch = [];
    callback();
    const commands = this.currentBatch;
    this.currentBatch = null;
    this.dispatch(doc => {
      commands.forEach(command => command(doc));
    });
  };

  markTouched(_path: FieldPath) {}

  getError(path: FieldPath) {
    return get(this.state.errors, path);
  }

  setError(path: FieldPath, message: string) {
    this.setState(prevState => ({
      ...prevState,
      errors: set(prevState.errors, path, message)
    }));
  }

  setErrors(errors: SuperFormErrors<T>) {
    this.setState({ errors });
  }

  arrayHelpers(path: FieldPath) {
    return new ArrayHelpers(path, this.dispatch);
  }

  canUndo() {
    return Automerge.canUndo(this.state.doc);
  }

  canRedo() {
    return Automerge.canRedo(this.state.doc);
  }

  undo() {
    this.setState(
      prevState => ({ doc: Automerge.undo(prevState.doc) }),
      () => {
        this.props.onChange && this.props.onChange(this.state.doc, this);
      }
    );
  }

  redo() {
    this.setState(
      prevState => ({ doc: Automerge.redo(prevState.doc) }),
      () => {
        this.props.onChange && this.props.onChange(this.state.doc, this);
      }
    );
  }

  render() {
    return (
      <SuperFormContext.Provider value={this}>
        <form
          onSubmit={e => {
            e.preventDefault();
            this.props.onSubmit && this.props.onSubmit(this.doc, this);
          }}
        >
          {this.props.children(this)}
        </form>
      </SuperFormContext.Provider>
    );
  }
}
