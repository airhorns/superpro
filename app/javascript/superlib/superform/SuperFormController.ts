import { get, set, toPath } from "lodash";
import { FieldPath, DocType } from "./utils";
import { ArrayHelpers } from "./ArrayHelpers";
import { Backend, AutomergeBackend, Command } from "./Backends";

export type SuperFormErrors<T extends DocType> = {
  [K in keyof T]?: T[K] extends object ? SuperFormErrors<T[K]> : string;
};

export type SuperFormChangeCallback<T extends DocType> = (newDoc: T, previousDoc: T, controller: SuperFormController<T>) => void;

export class SuperFormController<T extends DocType> {
  currentBatch: Command<T>[] | null = null;
  backend: Backend<T>;
  errors: SuperFormErrors<T> = {};
  clock = 0;

  constructor(initialDoc: T, onChange: SuperFormChangeCallback<T>, backendClass: typeof Backend = AutomergeBackend) {
    // Initialize the backend
    this.backend = new (backendClass as any)((doc: T, prevDoc: T) => {
      console.debug("doc", doc);
      console.debug("changes", this.backend.getChanges(prevDoc, doc));
      onChange(doc, prevDoc, this);
    });

    this.backend.resetDoc(initialDoc);
  }

  get doc(): T {
    return this.backend.doc as T;
  }

  getValue(path: FieldPath) {
    return get(this.backend.doc, path);
  }

  setValue(path: FieldPath, value: any) {
    this.dispatch(doc => set(doc, path, value));
  }

  deletePath(path: FieldPath) {
    const keys = toPath(path);
    this.dispatch(doc => {
      const root = get(doc, keys.slice(0, keys.length - 1));
      delete root[keys[keys.length - 1]];
    });
  }

  dispatch = (command: Command<T>) => {
    // Queue up changes for the batch if batching is underway
    if (this.currentBatch) {
      this.currentBatch.push(command);
      return;
    }

    // Otherwise, apply the change using the backend. The backend will call the onChange callback passed in the constructor.
    this.backend.change(command);
    this.clock += 1;
  };

  batch = (batchDefinition: () => void) => {
    if (this.currentBatch) {
      // This is probably possible but it likely indicates the calling code isn't factored right, nested batches adds overhead and
      // commands should operate either at the batch level or atomic level, and batch level ones should only call atomic ones.
      throw new Error("can't nest superform batches");
    }

    this.currentBatch = [];
    batchDefinition();
    const commands = this.currentBatch;
    this.currentBatch = null;
    this.dispatch(doc => {
      commands.forEach(command => command(doc));
    });
  };

  markTouched(_path: FieldPath) {
    return;
  }

  getError(path: FieldPath) {
    return get(this.errors, path);
  }

  setError(path: FieldPath, message: string) {
    set(this.errors, path, message);
  }

  setErrors(errors: SuperFormErrors<T>) {
    Object.assign(this.errors, errors);
  }

  arrayHelpers(path: FieldPath) {
    return new ArrayHelpers(path, this.dispatch);
  }

  canUndo() {
    return this.backend.canUndo();
  }

  canRedo() {
    return this.backend.canRedo();
  }

  undo() {
    return this.backend.undo();
  }

  redo() {
    return this.backend.redo();
  }
}
