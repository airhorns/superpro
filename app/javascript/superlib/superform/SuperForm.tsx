import { get, set, toPath } from "lodash";
import React from "react";
import { FieldPath, DocType } from "./utils";
import { ArrayHelpers } from "./ArrayHelpers";
import { SuperFormContext } from "../superform";
import { Backend, AutomergeBackend, Command } from "./Backends";
import { Box } from "grommet";

export interface SuperFormProps<T extends DocType> {
  children: (form: SuperForm<T>) => React.ReactNode;
  initialValues?: T;
  onChange?: (doc: T, form: SuperForm<T>) => void;
  onSubmit?: (doc: T, form: SuperForm<T>) => void;
  backendClass?: typeof Backend;
}

export type SuperFormErrors<T extends DocType> = {
  [K in keyof T]?: T[K] extends object ? SuperFormErrors<T[K]> : string;
};

interface SuperFormState<T extends DocType> {
  doc: T;
  errors: SuperFormErrors<T>;
}

export class SuperForm<T extends DocType> extends React.Component<SuperFormProps<T>, SuperFormState<T>> {
  currentBatch: Command<T>[] | null = null;
  backend: Backend<T>;

  constructor(props: SuperFormProps<T>) {
    super(props);

    // Initialize the backend
    const backendClass: any = props.backendClass || AutomergeBackend;
    this.backend = new backendClass((doc: T) => {
      const prevDoc = this.state.doc;
      this.setState({ doc }, () => {
        console.debug("doc", this.state.doc);
        console.debug("changes", this.backend.getChanges(prevDoc, this.state.doc));
        this.props.onChange && this.props.onChange(this.state.doc, this);
      });
    });

    if (this.props.initialValues) {
      this.backend.setInitialValues(this.props.initialValues);
    }

    this.state = { doc: this.backend.doc, errors: {} };
  }

  get doc(): T {
    return this.state.doc;
  }

  getValue(path: FieldPath) {
    return get(this.state.doc, path);
  }

  setValue(path: FieldPath, value: any) {
    this.dispatch(doc => set(doc, path, value));
  }

  deletePath(path: FieldPath) {
    const keys = toPath(path);
    this.dispatch(doc => {
      const root = get(doc, keys.slice(0, keys.length - 1));
      delete root[keys[keys.length - 1]];
    });
  }

  dispatch = (command: Command<T>) => {
    // Queue up changes for the batch if batching is underway
    if (this.currentBatch) {
      this.currentBatch.push(command);
      return;
    }

    // Otherwise, apply the change using the backend. The backend will call setState via it's onChange handler.
    this.backend.change(command);
  };

  batch = (callback: () => void) => {
    if (this.currentBatch) {
      // This is probably possible but it likely indicates the calling code isn't factored right, nested batches adds overhead and
      // commands should operate either at the batch level or atomic level, and batch level ones should only call atomic ones.
      throw new Error("can't nest superform batches");
    }

    this.currentBatch = [];
    callback();
    const commands = this.currentBatch;
    this.currentBatch = null;
    this.dispatch(doc => {
      commands.forEach(command => command(doc));
    });
  };

  markTouched(_path: FieldPath) {}

  getError(path: FieldPath) {
    return get(this.state.errors, path);
  }

  setError(path: FieldPath, message: string) {
    this.setState(prevState => ({
      ...prevState,
      errors: set(prevState.errors, path, message)
    }));
  }

  setErrors(errors: SuperFormErrors<T>) {
    this.setState({ errors });
  }

  arrayHelpers(path: FieldPath) {
    return new ArrayHelpers(path, this.dispatch);
  }

  canUndo() {
    return this.backend.canUndo();
  }

  canRedo() {
    return this.backend.canRedo();
  }

  undo() {
    return this.backend.undo();
  }

  redo() {
    return this.backend.redo();
  }

  render() {
    return (
      <SuperFormContext.Provider value={this}>
        <Box
          as="form"
          flex
          className="SuperForm"
          onSubmit={e => {
            e.preventDefault();
            this.props.onSubmit && this.props.onSubmit(this.doc, this);
          }}
        >
          {this.props.children(this)}
        </Box>
      </SuperFormContext.Provider>
    );
  }
}
