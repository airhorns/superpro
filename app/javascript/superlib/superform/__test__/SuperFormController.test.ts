import { SuperFormController } from "../SuperFormController";
import { SuperForm } from "../SuperForm";

interface TestDocShape {
  foo: boolean;
  list: number[];
  obj: { bar: boolean };
}
let controller!: SuperFormController<TestDocShape>;
let onChange!: jest.Mock<any, any>;

beforeEach(() => {
  onChange = jest.fn();
  controller = new SuperFormController<TestDocShape>({ foo: true, list: [1, 2, 3], obj: { bar: false } }, onChange);
});

test("it should allow values to be set at a path", () => {
  controller.setValue("foo", false);
  expect(controller.getValue("foo")).toBe(false);

  controller.setValue("obj.bar", true);
  expect(controller.getValue("obj.bar")).toBe(true);
});

test("it should call the change callback with the new doc when set", () => {
  controller.setValue("foo", false);
  expect(onChange).toMatchSnapshot();
});
