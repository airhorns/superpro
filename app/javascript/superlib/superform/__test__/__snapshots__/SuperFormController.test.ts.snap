// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should call the change callback with the new doc when set 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "foo": false,
        "list": Array [
          1,
          2,
          3,
        ],
        "obj": Object {
          "bar": false,
        },
      },
      Object {
        "foo": true,
        "list": Array [
          1,
          2,
          3,
        ],
        "obj": Object {
          "bar": false,
        },
      },
      SuperFormController {
        "backend": AutomergeBackend {
          "doc": Object {
            "foo": false,
            "list": Array [
              1,
              2,
              3,
            ],
            "obj": Object {
              "bar": false,
            },
          },
          "onChange": [Function],
        },
        "batch": [Function],
        "clock": 1,
        "currentBatch": null,
        "dispatch": [Function],
        "errors": Object {},
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
